### PR TITLE
[Perf] Enable CUDA graphs for the MammothModa2 AR stage

### PR DESCRIPTION
**Pull request by jyizheng**
Enable torch.compile piecewise + CUDA graphs for the AR stage of the MammothModa2 t2i deploy config. The AR stage dominates t2i latency (63.1s of 72.6s E2E on 1×H100, 4161 tokens).

Results (1×H100, 1024², 50 steps, greedy seed 42): AR 63.1s → 39.2s (**1.61×**), E2E 72.6s → 48.6s (**1.49×**), +0.17 GiB graph memory. Two same-config runs produce identical token sequences and bitwise-identical AR→DiT hidden states; images are quality-equivalent to eager.

`cudagraph_mode: FULL` without compilation stays unsupported — the dual-expert routing bakes the dummy all-text mask into the capture and t2i decode replays the wrong expert (#7318).

Refs #7075. Coordinated scope: torch.compile enablement itself is @Xenoryn's item; this PR only flips the deploy default and documents the graph-mode constraint.

---

**Update:** enabling graphs surfaced two pre-existing batch ≥ 8 correctness bugs, both fixed here (commit 6333a3d): (1) the dual-vocab embedding's host-branch + scatter path gathered the async-scheduling -1 sentinel — rewritten shape-static (dual gather + select, numerics-neutral: bitwise-identical outputs); `async_scheduling` is now explicitly off for this config since embed-early models are not async-safe. (2) stop-token finishes emitted the cumulative latent snapshot twice and the LATENT accumulation concatenated them — `ar2dit` now keeps the trailing snapshot. Validation: t2i batch 1/4/8/16 all pass (previously ≥ 8 crashed); bs1 AR 38.7s vs 63.1s eager (1.63×), E2E 1.49×; same-config runs bitwise-reproducible.

---

### Review follow-up (7f6f1cc)
- Trim removed; source fix: a stop-token finish delivers a full cumulative snapshot on top of the per-step chunks — the accumulator now supersedes chunks instead of appending it. Assert restored; unit + E2E (bs8 sampling, early stop-finishes) pass. Regression tests added.
- `cudagraph_mode` pinned. B300, same GPU/seed: eager 55.6s, PIECEWISE 109.1s (below eager), FULL_AND_PIECEWISE **26.6s (2.09×)**.
- Side-by-side t2i (identical inputs, greedy seed 42):

| eager (base) | this branch |
|---|---|
| eager | pr |

- I2T: responses word-for-word identical (config untouched; embedding rewrite bitwise-neutral).

### CODE REVIEW

**Comment by jyizheng** [comment]
Update from wider shape coverage: with this config, concurrent batches ≥ 8 hit a CUDA device-side assert in the dual-expert routing (`layer(..., gen_token_mask)`) during AR decode — bs1/bs4 are fine (1.51×/1.54×), and capture sizes do cover 8/16, so this is a routing graph-safety issue (same family as #7318), not a capture-coverage gap. Converting to draft while I root-cause with `CUDA_LAUNCH_BLOCKING=1`; will either fix the routing or scope the config change with a documented concurrency guard before marking ready.

**Comment by jyizheng** [comment]
Further isolation: the batch ≥ 8 crash reproduces with compilation only (`mode: 3, cudagraph_mode: NONE`) — a `vectorized_gather_kernel` index-out-of-bounds during early prefill/decode — so it is an inductor compilation issue in the dual-expert routing, not a CUDA-graph one. bs1/bs4 are unaffected. @Xenoryn this overlaps your "remove routing graph breaks" item; I'll take a stab at making `moe_forward`'s routing shape-static (fixes #7318's capture hazard too) unless you're already on it.

**Comment by Xenoryn** [comment]
> Further isolation: the batch ≥ 8 crash reproduces with compilation only (`mode: 3, cudagraph_mode: NONE`) — a `vectorized_gather_kernel` index-out-of-bounds during early prefill/decode — so it is an inductor compilation issue in the dual-expert routing, not a CUDA-graph one. bs1/bs4 are unaffected. @Xenoryn this overlaps your "remove routing graph breaks" item; I'll take a stab at making `moe_forward`'s routing shape-static (fixes #7318's capture hazard too) unless you're already on it.

Hi @jyizheng, thanks for coordinating the scope here! I noticed that the batch ≥ 8 fix in this PR now overlaps with #7432 in mammoth_moda2.py, particularly around the graph-safe embedding path.

If this PR lands first, I’ll rebase #7432 on top of it and adapt my changes accordingly. If #7432 lands first, just a heads-up that this PR may need a small rebase/update in that area. Happy to coordinate whichever lands first!

File: vllm_omni/model_executor/stage_input_processors/mammoth_moda2.py [hunk 1]
```diff
@@ -82,6 +82,15 @@ def ar2dit(
...
         full_hidden_states = mm_output["latent"]
         hidden_total = int(full_hidden_states.shape[0])
+        expected_total = len(prompt_token_ids) + len(gen_token_ids)
+        if hidden_total > expected_total:
+            # The AR stage emits the full cumulative hidden states with every
+            # output. A request that finishes on a stop token emits twice (the
+            # EOS step and the final flush), and the generic LATENT
+            # accumulation concatenates the two snapshots. The final snapshot
+            # is the complete one, so keep only the trailing expected rows.
+            full_hidden_states = full_hidden_states[-expected_total:]
```
Comment (lishunyang12) [inline vllm_omni/model_executor/stage_input_processors/mammoth_moda2.py]: Add regression tests for mixed-vocabulary embeddings and duplicated terminal latent snapshots.

**Comment by Xenoryn** [comment]
Could you add side-by-side E2E T2I images and I2T responses to the PR body, comparing main with this branch using identical inputs/settings?

**Review by linyueqian — CHANGES_REQUESTED** [review]
Requesting changes at `f56cd15f`. The measured win is real and the shape-static dual-vocabulary embedding is the right way to get there, but two of the three changes are not safe as written, and both are inline with the exact mechanism: the `ar2dit` trailing-row trim repairs a symptom on the wrong end of the tensor and disables the assert that would have caught it, and flipping `enforce_eager` on a backbone whose `forward` caches per-request metadata in Python means full graph replay reuses stale constraints once a batch changes composition. The `async_scheduling: false` half of the yaml is correct and the key is real; the embedding comment overclaims about the `-1` sentinel and I left that as a suggestion.

Two coordination notes: #7134 deletes `ar2dit` and replaces the length check with a strict `ValueError` in `ar2diffusion`, and #7432 rewrites `get_input_embeddings` and the same yaml block with `cudagraph_mode: NONE`, so this PR conflicts with both and the graph-mode decision needs to be made once, together with #7432's author, rather than in two PRs that undo each other. Reviewed statically with a two-model cross-check that reached the truncation finding independently; fork head, no PR code executed. Happy to re-review quickly once the trim is gone and the graph mode is pinned to something that still runs the Python `forward`.

**Comment by jyizheng** [comment]
All three addressed in 7f6f1cc. (1) Trim removed; root cause was a stop-token finish delivering a full cumulative snapshot on top of the per-step chunks — the accumulator now supersedes chunks; assert restored, tests added. (2) Mode pinned, but PIECEWISE regresses below eager on the same GPU (109.1s vs 55.6s; FULL decode graphs give 26.6s, 2.09×). Our gates (same-config bitwise repro, mid-run composition changes at bs8/16) have not caught a stale-state violation — happy to turn a concrete composition sequence into a regression test, and to settle the mode together with #7432. (3) Comment softened; `async_scheduling: false` kept.

@Xenoryn side-by-side images + i2t comparison are now in the PR body (i2t word-for-word identical). For the mode decision: propose #7432 keeps `cudagraph_mode: NONE` and this PR pins FULL_AND_PIECEWISE, so the two do not overwrite each other.

File: vllm_omni/model_executor/stage_input_processors/mammoth_moda2.py (hunk 1 above)
Comment (jyizheng) [reply vllm_omni/model_executor/stage_input_processors/mammoth_moda2.py]: Added in 7f6f1cc: `test_dual_vocab_embedding.py` + `test_latent_accumulation.py`.

File: vllm_omni/outputs/output_processor.py [hunk 2]
```diff
@@ -132,6 +142,16 @@ def add_multimodal_tensor(self, payload: Any | None, mm_type: str | None) -> Non
...
                 replace_snapshot_keys(self.mm_accumulated, incoming)
+                # Latent emissions are per-step chunks, except that a
+                # stop-token finish additionally delivers the full cumulative
+                # snapshot. A payload at least as long as everything
+                # accumulated so far supersedes the chunks; appending it
+                # would double-count the hidden states.
+                if modality_key == "latent":
+                    inc_rows = _dim0_rows(incoming.tensors.get("latent"))
+                    acc_rows = _dim0_rows(self.mm_accumulated.tensors.get("latent"))
+                    if inc_rows and acc_rows and inc_rows >= acc_rows:
```
Comment (linyueqian) [inline vllm_omni/outputs/output_processor.py]: [important] Tensor length cannot tell a cumulative snapshot from a legitimate next chunk, and this branch runs for every latent producer (`moss_tts`, `audex`, `minimax_music3` and MammothModa2 all emit `engine_output_type="latent"`), not only this model. The simplest trigger is not chunked prefill: a one-token prompt accumulates one row, the first decode step delivers one row, `1 >= 1` holds, and the prefill latent is popped, so the request finishes without its prompt history. Two successive 32-row prefill slices under chunked prefill fail the same way. Either case then trips the `hidden_total != len(full_token_ids)` check in `stage_input_processors/mammoth_moda2.py`. Mark the stop-token snapshot explicitly where it is produced (or at the transport boundary) and keep plain concatenation for ordinary chunks; add one-row-then-one-row, equal-sized and growing-chunk cases next to the terminal-snapshot test so the heuristic cannot come back.

File: vllm_omni/deploy/mammoth_moda2.yaml [hunk 3]
```diff
@@ -9,7 +9,21 @@ stages:
...
+    # torch.compile piecewise + CUDA graphs (vLLM's FULL_AND_PIECEWISE
+    # default, pinned explicitly). PIECEWISE alone regresses below eager
+    # (compiled kernels without full-graph launch amortization); the win
+    # comes from the FULL decode graphs. Do not use cudagraph_mode FULL
+    # without compilation: the dual-expert (und/gen) routing would be
+    # captured with the dummy all-text mask and replay the wrong expert
+    # during t2i decode (#7318).
+    enforce_eager: false
+    compilation_config:
+      cudagraph_mode: FULL_AND_PIECEWISE
```
Comment (linyueqian) [inline vllm_omni/deploy/mammoth_moda2.yaml:21]: [important] `FULL_AND_PIECEWISE` still replays FULL decode graphs, and a replayed decode skips the Python `forward` that assigns `_last_runtime_additional_information`, so the round-one stale-metadata finding stands. The runner updates `generated_len` inside the referenced dicts but builds a new list in current batch order every step; the model's cached list does not follow that order. Concretely: prefill [A, B] with different grid widths or generation offsets, finish A, then decode B in slot 0 without another prefill; batch condensation moves B but `compute_logits` still reads A's metadata at index 0 and applies the wrong EOL constraint. Refresh that metadata outside the captured forward on every step, or keep a mode that executes the Python path, and pin this sequence in a regression test. The 2.09x from FULL decode graphs is real, but not at the cost of wrong constraints under reordering.

File: tests/engine/test_latent_accumulation.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+    def __init__(self):
+        self.mm_type = None
+        self.mm_accumulated = MultimodalPayload()
+
+    def add(self, tensor):
+        OmniRequestState.add_multimodal_tensor(self, {"latent": tensor}, "latent")
+
+    def consolidated(self):
+        self.mm_accumulated.consolidate_tensors(get_accumulation_strategy(OutputModality.LATENT))
```
Comment (linyueqian) [inline tests/engine/test_latent_accumulation.py]: [important] `consolidate_tensors` takes an `OutputModality` and resolves the strategy internally; this helper passes `TensorAccumulationStrategy.CONCAT_DIM0`, so `test_per_step_chunks_concatenate` hits `OutputModality.AUDIO in modality` on an enum member and raises `TypeError` before it asserts anything. The snapshot tests do not notice because replacement leaves a single tensor and never reaches the concat path, so the concatenation this PR claims to cover is currently untested. The file carries `core_model` and `cpu` marks, so the general lane will be red at this head once it runs. Pass `OutputModality.LATENT` here.

**Review by linyueqian — CHANGES_REQUESTED** [review]
Re-reviewed at `fd62db93` against the round-one head `f56cd15f`. Two of the three round-one items are closed: the `ar2dit` trailing-row trim is gone (the head now carries the #7134 `ar2diffusion` migration and keeps its strict length-mismatch `ValueError`), and the embedding comment no longer claims safety during async input preparation while `async_scheduling: false` stays. The graph-mode item is not closed by pinning `FULL_AND_PIECEWISE`, and the replacement for the trim moved the problem into the shared latent accumulator; both are inline, together with a broken helper in the new test. The description still carries the obsolete "Update" paragraph about the `ar2dit` trim, so please refresh it to describe the accumulator change.

The FULL versus PIECEWISE decision still needs to be made once with #7432, which is under review in parallel; until the per-request metadata is refreshed outside the captured forward, a mode that replays FULL decode graphs is not safe for this backbone. Two independent reviewers confirmed all three items and supplied the one-row trigger for the accumulator; one minor note from them, not a finding: the graph-safe `torch.where` embedding rewrite now runs both embedding lookups for every token, a small steady-state cost on the hot path that is a fair trade for graph safety. Static read, both heads against their own merge-bases; fork head, no PR code executed; pre-commit and DCO green, no general lane has run at this head.

**Comment by jyizheng** [comment]
All three in 2d6ad24 (+ description refreshed). (1) Heuristic replaced by prefix-bitwise-containment — lossless by construction; your three counterexamples are pinned as tests. (2) Took your first option: per-step metadata refresh outside capture, condensation sequence pinned as a test. Does that close the mode item for FULL_AND_PIECEWISE (coordinating with #7432)? (3) Helper fixed. 15 unit tests + bs8 stop-finish E2E pass.
